### ISSUE TITLE
Implement the boolean NOT operator in Ash, rather than in the compiler

### ISSUE THREAD

**Issue by SamTebbs33** [ash-lang/ash#45]
With the new operator definition system, it would be possible to define the "!" as just a prefix operator, and define its behaviour through operator overloads. However, this depends on the implementation of #44

```
operator ! = type: prefix, name: bool_not

// The below function would generate the same code that is used by a "!" operator in Java 
prefix func !(operand : bool) : bool -> operand ? false : true
```

If this proves to be too cumbersome, then the "!" operator could be handled natively by the compiler, although this is not preferable.

**Comment by SamTebbs33** [comment]
It's been decided that implementing the optional unwrap behaviour through an operator overload would be too cumbersome, error-prone (the operand would need to be casted from `Object?` to `Object`, which is unsafe in all use cases) and would allow non-optionals to be unwrapped (as a non-optional can be passed to a function as an optional), which is semantically invalid.

**Comment by SamTebbs33** [comment]
Implemented in dd01527 and b571d80
